vmdb/app/models/metric/ci_mixin/capture/openstack_base.rb

```diff
@@ -37,134 +37,355 @@ def perf_release_openstack
     counter_values_by_ts = {}
     metric_capture_module::COUNTER_INFO.each do |i|
       timestamps = i[:openstack_counters].collect { |c| metrics_by_counter_name[c].try(:keys) }
       timestamps = timestamps.flatten.compact.uniq.sort
 
-      aggregate = []
-      agg_start = nil
-      timestamps.each_cons(2) do |last_ts, ts|
-        agg_start ||= last_ts
+      if i[:openstack_counters].count == 1
+        $log.debug "#{log_header} id:[#{name}] started collecting single counter stats for: "\
+                   "#{i[:openstack_counters]}, with available data on timestamps: #{timestamps}"
+        process_single_counter_stats!(counter_values_by_ts, metric_capture_module, i, timestamps,
+                                      metrics_by_counter_name)
+      else
+        $log.debug "#{log_header} id:[#{name}] started collecting multi counter stats for: "\
+                   "#{i[:openstack_counters]}, with available data on timestamps: #{timestamps}"
+        process_multi_counter_stats!(counter_values_by_ts, metric_capture_module, i, timestamps,
+                                     metrics_by_counter_name, data_collecting_period, log_header)
+      end
+    end
 
-        interval     = ts - last_ts
-        metrics      = i[:openstack_counters].collect { |c| metrics_by_counter_name.fetch_path(c, ts) }
-        value        = i[:calculation].call(*metrics, interval)
+    counter_values_by_ts
+  end
 
-        # break down values from cumulative meters into discrete values
-        if metric_capture_module.diff_meter? i[:openstack_counters]
-          last_metrics = i[:openstack_counters].collect { |c| metrics_by_counter_name.fetch_path(c, last_ts) }
-          last_value   = i[:calculation].call(*last_metrics, interval)
-          value -= last_value
-        end
+  def process_single_counter_stats!(counter_values_by_ts, metric_capture_module, i, timestamps, metrics_by_counter_name)
+    # !!! This method modifies counter_values_by_ts
+    # We have only one counter, we can pass the values and intervals to calculation in simplified manner, which
+    # is compatible with previous version. We are passing values into calculation methods, not hashes.
+
+    timestamps.each_cons(2) do |last_period, period|
+      metrics = {}
+      i[:openstack_counters].each { |c| metrics[c] = metrics_by_counter_name.fetch_path(c, period) }
+
+      if metric_capture_module.diff_meter? i[:openstack_counters]
+        last_metrics = {}
+        i[:openstack_counters].each { |c| last_metrics[c] = metrics_by_counter_name.fetch_path(c, last_period) }
+        # Interval is related only to diff metrics
+        intervals = {}
+        i[:openstack_counters].each { |c| intervals[c] = metrics[c][:duration_end] - last_metrics[c][:duration_end] }
+        diff_metrics_avg = {}
+        i[:openstack_counters].each { |c| diff_metrics_avg[c] = metrics[c][:avg] - last_metrics[c][:avg] }
```

Comment:
this will not fit one line  last_metrics = i[:openstack_counters].each_with_object({}) { |c, h| h[c] = metrics_by_counter_name.fetch_path(c, last_period) }

so I would have to use multi-line block and it would beeven more lines I guess :-)